base/shared/src/main/scala/scalaz/data/Freer.scala

```diff
@@ -0,0 +1,86 @@
+package scalaz
+package data
+
+import scala.annotation.tailrec
+
+/**
+  * Free(r) Monad - arrows A => Free[F, B] are concatenated using a type-aligned
+  * stack-ended queue ([[ACatenable1]]), for better complexity on certain types
+  * of Free monad use cases.
+  *
+  * For more information, [see Ed Kmett's comments here.](https://www.reddit.com/r/haskell/comments/7q4sku/are_people_using_freer_monads_or_still_mostly/dsmlnh7/)
+  *
+  */
+sealed abstract class Freer[F[_], A] {
+
+  def map[B](f: A => B): Freer[F, B] =
+    flatMap[B](a => Freer.pure[F, B](f(a)))
+
+  def flatMap[B](k: A => Freer[F, B]): Freer[F, B] =
+    this match {
+      case Freer.Pure(a) => k(a)
```

Comment:
This eagerness is tempting, but not stack-safe.

Consider

```scala
Pure(a).flatMap(x => Pure(x).flatMap(x => Pure(x).flatMap(x => Pure(x).flatMap(...))))
```

Reply:
That's definitely the next pass over this code - making sure everything is stacksafe. I'm noticing the implementation of `foldMap` in both `Free` and this one aren't stacksafe either.